src/cockpit/polkit.py

```diff
@@ -0,0 +1,140 @@
+                process.terminate()
+                break
+
+            logger.debug('Read line from helper: %s', line)
+            # usually: PAM_PROMPT_ECHO_OFF Password: \n
+            if line.startswith(b'PAM_PROMPT'):
+                prompt = line.split(b' ', 1)[1].strip().decode()
+                # Don't pass this to the UI if it's "Password:" (the usual case),
+                # so that superuser.py uses the translated default
+                if prompt.startswith('Password'):
+                    prompt = ''
+
+                try:
+                    passwd = await self.responder.do_askpass('', prompt, '')
+                    logger.debug('got PAM_PROMPT %s, do_askpass returned a password', prompt)
+                except asyncio.CancelledError:
+                    logger.debug('got PAM_PROMPT %s, do_askpass cancelled', prompt)
+                    process.terminate()
+                    break
+                process.stdin.write(passwd.encode())
+                process.stdin.write(b'\n')
+                del passwd  # don't keep this around longer than necessary
+                await process.stdin.drain()
+                logger.debug('got PAM_PROMPT, wrote password to helper')
+            elif line.startswith(b'PAM_TEXT_INFO'):
+                logger.info(line.decode())
+            elif line.startswith(b'PAM_ERROR'):
+                logger.warning(line.decode())
+            elif line.startswith(b'SUCCESS'):
+                logger.info('Authentication succeeded')
+                break
+            elif line.startswith(b'FAILURE'):
+                logger.warning('Authentication failed')
+                break
+            else:
+                logger.warning('Unknown line from helper, aborting: %s', line)
+                process.terminate()
+                break
+
+        res = await process.wait()
+        logger.debug('helper exited with code %i', res)
+
+    @bus.Interface.Method('', 's')
+    def cancel_authentication(self, cookie):
+        logger.warning('CancelAuthentication: %s not implemented', cookie)
+
+
+def register_agent(responder):
+    '''Register polkit agent.
+
+    Returns an opaque object which needs to be kept during the lifetime of the agent.
+    After that, pass it to @unregister_agent().
+    '''
+    try:
+        system_bus = bus.Bus.default_system()
+    except FileNotFoundError:
+        logger.debug('system bus not found, not registering polkit agent')
+        return None
+
+    try:
+        # may refine that with a D-Bus call to logind
+        subject = ('unix-session', {'session-id': {'t': 's', 'v': os.environ['XDG_SESSION_ID']}})
+    except KeyError:
+        logger.debug('XDG_SESSION_ID not set, not registering polkit agent')
+        return None
+
+    agent_object = org_freedesktop_PolicyKit1_AuthenticationAgent(responder)
+    agent_slot = system_bus.add_object('/', agent_object)
+    logger.debug('added agent object %s', agent_slot)
```

Comment:
Debugging leftover, we don't need to print agent_slot or that whole message really.

Reply:
Removed.